File: runners/elixir/verify_pqc_mldsa.exs
# Elixir runner for the post-quantum ML-DSA-65 corpus (pqc_mldsa_v0).
#
# Independent Elixir port of the Python reference runner
# (runners/python/verify_pqc_mldsa.py) and its decision surface
# (tools/oracle_pqc_mldsa.py): decode the hex public key, message and signature,
# reject a wrong-length public key (must be 1952) or signature (must be 3309)
# before any verify, then verify the FIPS-204 ML-DSA-65 signature over the exact
# message bytes with the EMPTY context string (the pure ML-DSA variant).
#
# Erlang/OTP's :crypto is built against an OpenSSL that predates ML-DSA (FIPS 204
# landed in OpenSSL 3.5; the OTP-26 image ships OpenSSL 3.0.20, whose public-key
# set has no ML-DSA), OTP has no built-in FFI, and there is no mature FIPS-204
# NIF on hex. So this runner keeps the ENTIRE decision surface in the BEAM (hex
# decode, wrong-length rejection) and shells out through an Erlang port
# (System.cmd) to a tiny C helper (mldsa_verify_helper.c) for the ONE primitive
# it cannot do in-VM: the raw ML-DSA-65 verify. The helper links liboqs, the same
# C reference the C runner uses; OQS_SIG_new returns NULL on an old Dilithium-only
# build, which the helper surfaces as exit 2 (the built-in tripwire).
#
# The helper is compiled on first run (cc ... -loqs) next to this script, so a
# single `elixir verify_pqc_mldsa.exs corpus.json` works given liboqs + a C
# compiler. Requires jason for JSON.
#
# Corpus path: argv[0], else the repo corpus relative to this source directory.
#
# Run:  cd runners/elixir && elixir verify_pqc_mldsa.exs [corpus.json]

Mix.install([{:jason, "~> 1.4"}])

defmodule VerifyPqcMldsa do
  @default_path "../../corpus/pqc_mldsa_v0/pqc_mldsa_v0.json"
  @sections ~w(mldsa65_verify mldsa65_malformed mldsa65_acvp_kat)

  @pk_len 1952
  @sig_len 3309

  # hex -> binary; :error on odd length or a bad digit. An empty string decodes
  # to a real 0-byte value, so the empty-message case is a real 0-byte message.
  defp hexdec(s) when is_binary(s) do
    if rem(byte_size(s), 2) == 0 and String.match?(s, ~r/\A[0-9a-fA-F]*\z/) do
      {:ok, Base.decode16!(s, case: :mixed)}
    else
      :error
    end
  end

  defp hexdec(_), do: :error

  # The BEAM-side decision surface: reject a malformed public key / signature
  # (wrong byte length, empty included) before any verify, exactly as the Python
  # oracle does. Only a well-formed, correctly sized triple reaches the helper.
  defp verdict(helper, pk_hex, msg_hex, sig_hex) do
    with {:ok, pk} <- hexdec(pk_hex),
         {:ok, _msg} <- hexdec(msg_hex),
         {:ok, sig} <- hexdec(sig_hex),
         true <- byte_size(pk) == @pk_len,
         true <- byte_size(sig) == @sig_len do
      # Empty message stays an empty argv element (a real 0-byte message).
      {_out, status} =
        System.cmd(helper, [pk_hex, msg_hex, sig_hex], stderr_to_stdout: true)

      case status do
        0 -> true
        1 -> false
        _ -> raise "mldsa_verify_helper failed (exit #{status}); is liboqs ML-DSA-65 present?"
      end
    else
      _ -> false
    end
  end

  # Compile the liboqs C helper next to this script on first run; reuse it after.
  defp ensure_helper(dir) do
    bin = Path.join(dir, "mldsa_verify_helper")
    src = Path.join(dir, "mldsa_verify_helper.c")

    unless File.exists?(bin) do
      {out, status} =
        System.cmd("cc", ["-O2", "-w", "-o", bin, src, "-loqs"], stderr_to_stdout: true)

      if status != 0 do
        IO.puts(:stderr, "failed to build mldsa_verify_helper:\n#{out}")
        System.halt(2)
      end
    end

    bin
  end

  def main(argv) do
    dir = __DIR__
    path = List.first(argv) || Path.join(dir, @default_path)
    helper = ensure_helper(dir)
    corpus = Jason.decode!(File.read!(path))

    results =
      Enum.flat_map(@sections, fn sec ->
        Enum.map(corpus[sec] || [], fn c ->
          accept = verdict(helper, c["public_key"], c["message"], c["signature"])
          {accept == c["expect_valid"], sec, c["note"]}
        end)
      end)

    fails = Enum.filter(results, fn {ok, _, _} -> not ok end)
    Enum.each(fails, fn {_, s, n} -> IO.puts("FAIL  [#{s}] #{n}") end)
    matched = Enum.count(results, fn {ok, _, _} -> ok end)
    total = length(results)
    IO.puts("\nelixir (pqc_mldsa): #{matched}/#{total} cases matched")
    System.halt(if length(fails) == 0, do: 0, else: 1)
  end
end

VerifyPqcMldsa.main(System.argv())

File: runners/c/verify_pqc_mldsa.c
/* C runner for the post-quantum ML-DSA-65 corpus (pqc_mldsa_v0).
 *
 * Independent C port of the Python reference runner (runners/python/verify_pqc_mldsa.py)
 * and its decision surface (tools/oracle_pqc_mldsa.py): decode the hex public key,
 * message and signature, reject a wrong-length public key (must be 1952) or
 * signature (must be 3309) before any verify, then verify the FIPS-204 ML-DSA-65
 * signature over the exact message bytes with the EMPTY context string.
 *
 * The ML-DSA implementation is liboqs via its C API (OQS_SIG_new("ML-DSA-65") /
 * OQS_SIG_verify). OQS_SIG_verify verifies the pure ML-DSA variant (empty context).
 * OQS_SIG_new returns NULL if the installed liboqs is an old build that exposes
 * only round-3 "Dilithium3" and not "ML-DSA-65"; that is the built-in tripwire.
 * JSON via jansson.
 *
 * Build (matches kaf/run_cells_pqc_mldsa.sh, which first builds liboqs 0.14.0):
 *   cc -O2 -w -o verify_pqc_mldsa verify_pqc_mldsa.c $(pkg-config --cflags --libs jansson) -loqs -lcrypto
 * Corpus path: argv[1], else $ALGOVOI_PQC_MLDSA, else ../../corpus/pqc_mldsa_v0/pqc_mldsa_v0.json
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <jansson.h>
#include <oqs/oqs.h>

#define PK_LEN 1952
#define SIG_LEN 3309

static const char *jstr(json_t *o, const char *k){
    json_t *v = json_object_get(o, k);
    return (v && json_is_string(v)) ? json_string_value(v) : NULL;
}

static int nib(char c){
    if(c>='0'&&c<='9') return c-'0';
    if(c>='a'&&c<='f') return c-'a'+10;
    if(c>='A'&&c<='F') return c-'A'+10;
    return -1;
}

/* hex decode into a malloc'd buffer; NULL on odd length or a bad digit. An empty
 * string decodes to a zero-length buffer (non-NULL), so the empty-message case is
 * a real 0-byte message, not a decode failure. */
static unsigned char *hexdec(const char *s, size_t *outlen){
    if(!s) return NULL;
    size_t n = strlen(s);
    if(n % 2) return NULL;
    size_t bl = n / 2;
    unsigned char *b = malloc(bl ? bl : 1);
    if(!b) return NULL;
    for(size_t i=0;i<bl;i++){
        int hi = nib(s[2*i]), lo = nib(s[2*i+1]);
        if(hi<0 || lo<0){ free(b); return NULL; }
        b[i] = (unsigned char)((hi<<4)|lo);
    }
    *outlen = bl;
    return b;
}

static int verdict(OQS_SIG *sig, const char *pkHex, const char *msgHex, const char *sigHex){
    size_t pl=0, ml=0, sl=0;
    int rc = 0;
    unsigned char *pk = hexdec(pkHex, &pl);
    unsigned char *msg = hexdec(msgHex, &ml);
    unsigned char *s = hexdec(sigHex, &sl);
    if(pk && msg && s && pl == PK_LEN && sl == SIG_LEN){
        rc = (OQS_SIG_verify(sig, msg, ml, s, sl, pk) == OQS_SUCCESS) ? 1 : 0;
    }
    free(pk); free(msg); free(s);
    return rc;
}

int main(int argc, char **argv){
    const char *env = getenv("ALGOVOI_PQC_MLDSA");
    const char *path = argc > 1 ? argv[1]
                     : (env ? env : "../../corpus/pqc_mldsa_v0/pqc_mldsa_v0.json");
    json_error_t err;
    json_t *corpus = json_load_file(path, 0, &err);
    if(!corpus){ fprintf(stderr, "cannot load corpus %s: %s\n", path, err.text); return 2; }

    OQS_SIG *sig = OQS_SIG_new("ML-DSA-65");
    if(!sig){ fprintf(stderr, "liboqs has no ML-DSA-65 (old/Dilithium build)\n"); json_decref(corpus); return 2; }

    const char *sections[] = {"mldsa65_verify", "mldsa65_malformed", "mldsa65_acvp_kat"};
    int total = 0, matched = 0;
    for(size_t si = 0; si < sizeof(sections)/sizeof(sections[0]); si++){
        json_t *sec = json_object_get(corpus, sections[si]);
        size_t i; json_t *c;
        json_array_foreach(sec, i, c){
            const char *pk = jstr(c, "public_key");
            const char *msg = jstr(c, "message");
            const char *s = jstr(c, "signature");
            const char *note = jstr(c, "note");
            int expect = json_is_true(json_object_get(c, "expect_valid"));
            int accept = verdict(sig, pk, msg, s);
            total++;
            if(accept == expect) matched++;
            else printf("FAIL  [%s] %s\n", sections[si], note ? note : "");
        }
    }
    printf("\nc (pqc_mldsa): %d/%d cases matched\n", matched, total);
    OQS_SIG_free(sig);
    json_decref(corpus);
    return matched == total ? 0 : 1;
}
